lms/static/scripts/frontend_apps/utils/api.ts

```diff
@@ -209,3 +212,75 @@ export function useAPIFetch<T = unknown>(
   const paramStr = recordToQueryString(params ?? {});
   return useFetch(path ? `${path}${paramStr}` : null, fetcher);
 }
+
+export type PaginatedFetchResult<T> = Omit<FetchResult<T>, 'mutate'> & {
+  /** Determines if currently loading the first page */
+  isInitiallyLoading: boolean;
+  loadNextPage: () => void;
+};
+
+/**
+ * Hook that fetches paginated data using authenticated API requests.
+ * It expects the result to include pagination info in a `pagination` prop.
+ *
+ * @param prop - Prop containing the list of results to collect.
+ * @param path - Path for API call, or null if there is nothing to fetch
+ * @param [params] - Query params for API call
+ */
+export function usePaginatedAPIFetch<
+  Prop extends string,
+  ListType extends any[],
+  FetchType extends { [P in Prop]: ListType } & { pagination: Pagination },
+>(
+  prop: Prop,
+  path: string | null,
+  params?: Record<string, string | string[] | undefined>,
+): PaginatedFetchResult<ListType> {
+  const [nextPageURL, setNextPageURL] = useState<string>();
+  const [currentList, setCurrentList] = useState<ListType>();
```

Comment:
We will be loading all loaded pages in an in-memory array. If the amount of pages is huge, this could have performance issues, or not be practical, but let's be simple for now.

We could replace this with some kind of iterator that discards oldest pages as newer ones are loaded, but let's postpone that until it becomes a clear problem.